### ISSUE TITLE
solid/reactivity false positive (since 0.16.1): capture read inside a synchronous array-method callback in createMemo/createEffect

### ISSUE THREAD

**Issue by niko86** [joshwilsonvu/eslint-plugin-solid#223]
**Since 0.16.1, `solid/reactivity` flags a capture that is read inside a *synchronous* array-method callback within a `createMemo`/`createEffect` body** — reporting "a returned function reads the capture later — it will never update", even though the callback runs during the tracked computation and the capture is refreshed on every re-run.

## Minimal reproduction

`eslint.config.mjs`:

```js
import solid from "eslint-plugin-solid/configs/recommended";

export default [
  {
    files: ["**/*.jsx"],
    ...solid,
    languageOptions: {
      ecmaVersion: 2024,
      sourceType: "module",
      parserOptions: { ecmaFeatures: { jsx: true } },
    },
  },
];
```

`repro.jsx`:

```jsx
import { createEffect, createMemo, createSignal } from "solid-js";

const ITEMS = ["alpha", "beta", "gamma"];

export function Filtered() {
  const [query, setQuery] = createSignal("");

  const matches = createMemo(() => {
    const q = query().toUpperCase();
    return ITEMS.filter((item) => item.toUpperCase().includes(q));
  });

  createEffect(() => {
    const q = query();
    console.log(ITEMS.some((item) => item === q));
  });

  return (
    <input
      value={query()}
      onInput={(e) => setQuery(e.currentTarget.value)}
      title={String(matches().length)}
    />
  );
}
```

`npx eslint repro.jsx`:

```
  13:15  warning  'q' captures the value of the reactive variable 'query' at setup, but a returned function reads the capture later — it will never update. Call 'query' inside the returned function instead, or prefix 'q' with 'initial'/'default'/'static' if a one-time snapshot is intended  solid/reactivity
  19:15  warning  'q' captures the value of the reactive variable 'query' at setup, but a returned function reads the capture later — it will never update. Call 'query' inside the returned function instead, or prefix 'q' with 'initial'/'default'/'static' if a one-time snapshot is intended  solid/reactivity
```

## Versions (A/B/A on the identical repro)

| eslint-plugin-solid | result |
| --- | --- |
| 0.16.0 | clean |
| 0.16.1 | 2 warnings (introduced here) |
| 0.17.0 | 2 warnings |

eslint v10.10.0 · node v24.14.1 · solid-js latest

## Why I believe these are false positives

- `Array.prototype.filter`/`.some` invoke their callbacks synchronously, inside the memo/effect body. Nothing "reads the capture later": by the time `createMemo` returns, every read already happened.
- The capture cannot go stale. When `query` changes, the enclosing tracked scope re-runs and re-captures — reactivity flows through the scope, not the variable.
- Hoisting `query().toUpperCase()` out of the per-item callback is the standard "compute once per run, not once per element" shape. The message's suggested fix (call the signal inside the callback) multiplies the work by the element count.
- The other escape hatch — prefixing with `initial`/`default`/`static` — mislabels the code: these are not one-time snapshots, they are per-run captures that *are* reactive via the enclosing scope.

## Impact

Taking 0.16.0 → 0.17.0 on a mid-sized SolidJS app flagged 19 sites across 10 files, essentially all of this shape (plus a couple of genuinely intended snapshots the prefix convention handles fine). We've pinned 0.16.0 for now.

If the analysis intends to treat array-method callbacks as escaping functions, a special case for the well-known synchronous iteration methods (as the pre-0.16.1 analysis appeared to have) would remove this whole class.

**Comment by ryansolid** [comment]
Fixed in [v0.17.1](https://github.com/solidjs-community/eslint-plugin-solid/releases/tag/v0.17.1), just published. The `staleCapture` check was treating any path from a capturing function to a `return` statement as an escape, but a function passed as a call argument (your `ITEMS.filter((item) => ...)`) doesn't escape through the return — only the call's result does, and the callback runs synchronously inside the computation where the capture is still fresh. Both repros from this issue are now regression tests. Thanks for the clear report!
